skimage/restoration/deconvolution.py

```diff
@@ -383,3 +383,96 @@ def richardson_lucy(image, psf, iterations=50, clip=True, filter_epsilon=None):
         im_deconv[im_deconv < -1] = -1
 
     return im_deconv
+
+
+def DNP_Gauss_freq(image, psf, we=0.01, clip=True):
+    """ Deconvolution using natural image priors
+        with a Gaussian prior, solved in the frequency domain
+
+    Parameters
+    ----------
+    image : ndarray
+       Input degraded image.
+    psf : ndarray
+       The point spread function.
+    we : float, optional
+       Smoothness weight, see [2]
+    clip : boolean, optional
+       True by default. If true, pixel value of the result above 1 or
+       under 0 are thresholded for skimage pipeline compatibility.
+
+    Returns
+    -------
+    im_deconv : ndarray
+       The deconvolved image.
+
+    Examples
+    --------
+    >>> from skimage import image_as_float, data, restoration
+    >>> camera = img_as_float(data.camera())
+    >>> from scipy.signal import convolve2d
+    >>> psf = np.ones((5, 5)) / 25
+    >>> camera = convolve2d(camera, psf, 'same')
+    >>> camera += 0.1 * camera.std() * np.random.standard_normal(camera.shape)
+    >>> deconvolved = restoration.DNP_Gauss_freq(camera, psf)
+
+    Notes
+    -----
+    This algorithm solves the deconvolution in the frequency domain. Artefacts
+    like 'ringing' on sharp transitions are strongly reduced in comparison to
+    other deconvolution strategies.
+    The smoothness weight needs to be estimated by trial and error. See [1-3]
+    for guidance.
+
+    References
+    ----------
+    .. [1] http://groups.csail.mit.edu/graphics/CodedAperture/
+    .. [2] Levin, A., Fergus, R., Durand, F., & Freeman, W. T. (2007).
+           Deconvolution using natural image priors.
+           Massachusetts Institute of Technology,
+           Computer Science and Artificial Intelligence Laboratory, 3.
+    .. [3] Levin, A., Fergus, R., Durand, F., & Freeman, W. T. (2007).
+           Image and depth from a conventional camera with a coded aperture.
+           ACM transactions on graphics (TOG), 26(3), 70-es.
+    """
+    float_type = np.promote_types(image.dtype, np.float32)
+    image = image.astype(float_type, copy=False)
+    psf = psf.astype(float_type, copy=False)
+
+    n, m = image.shape
+    # psf.shape() is expected to be odd in both dimension but works with even
+    # too.
+    # not sure if the double flip is needed
+    psf = np.fliplr(np.flipud(psf))
+
+    # force some shapes
+    onesrow = np.array([-1, 1])
+    onesrow.shape = (1, 2)
+    onescol = np.array([-1, 1])
+    onescol.shape = (2, 1)
+    Gx = np.fft.fft2(onesrow, s=(n, m))
+    Gy = np.fft.fft2(onescol, s=(n, m))
+    F = np.fft.fft2(psf, s=(n, m))
```

Comment:
Could we use fftn here in order to allow support for 3D+ images? Ditto for using `np.flip(..., axis=i)` instead of `fliplr` and `flipud`. I think with a small amount of effort this code can be generically n-dimensional.

Reply:
Please import the fft module from skimage._shared.fft as in the following file:
https://github.com/scikit-image/scikit-image/blob/c8e927e8d379943920486013c38376dedd3280de/skimage/filters/lpi_filter.py#L7
You can then call the functions using the functions from that module:
```Python
fft.fftn(onescol, s=(n, m))
```
This will result in using scipy.fft in place of numpy.fft whenver the user has SciPy >= 1.4. The SciPy implementation is preferable for three main reasons:
1.) it can perform transforms in single precision
2.) it can be configured to use multiple threads
3.) it can be overridden by alternative backends such as MKL or PyFFTW

Reply:
@grlee77 thanks, changed to skimage fft. ready in next commit.
@jni I still work on the conversion to N dimensions, which is not trivial. I will try to get the 3D going in the next commit

Reply:
For the implementation proposed in alexjarosch#1, the psf has the same number of dimensions as the image. I can also add multichannel support if we want to allow e.g. 2d psf for a 2d + channels image, etc.